OpenMarket/OpenMarket/Decoder.swift

```diff
@@ -7,19 +7,16 @@
 
 import UIKit
 
-
 struct Decoder {
     let decoder = JSONDecoder()
-
-    func parsePageJSON() throws -> ProductList {
-        guard let pageJSON: NSDataAsset = NSDataAsset(name: "products") else {
-            throw NetworkError.parsingFailed
-        }
-        
-        guard let decodedPageJSON = try decoder.decode(ProductList?.self, from: pageJSON.data) else {
-            throw NetworkError.parsingFailed
+    
+    func parsePageJSON (data: Data) -> ProductList? {
+        do {
+            let pageJSON: Data = data
+            let decodedPageJSON = try decoder.decode(ProductList.self, from: pageJSON)
+            return decodedPageJSON
+        } catch {
+            return nil
         }
```

Comment:
이미 아시겠지만 throwing function을 사용할 때 디코딩 실패 시 에러 발생 이유를 알리지 않고 단순히 `nil`을 반환하고자 한다면 optional try 문법을 사용해볼 수 있을 것 같아요.

```suggestion
        let pageJSON: Data = data
        guard let decodedPageJSON = try? decoder.decode(ProductList.self, from: pageJSON) else {
            return nil
        }
        return decodedPageJSON
```

기본 프레임워크에서 제공하는 메서드들은 대부분 에러 발생 시 발생 이유를 자세히 알려주기에 개인적으로는 메서드 사용자의 입장을 고려해서 에러 발생 원인을 알려주는 것이 좋다고 생각합니다.
처리 방법은 아래 중에서 선택할 수 있을 것 같아요. 호랭이는 어떤 것이 좋을 것 같은지, 아니면 다른 의견이 있는지 이야기 나누어보았으면 좋겠어요~!
1. 출력문 (로그)
2. 동일하게 throw 메서드로 구성하여 메서드 사용자에게 처리를 맡김
3. Result 타입을 통해 에러 반환
4. [컴파일러 전처리문](https://docs.swift.org/swift-book/ReferenceManual/Statements.html#ID538)을 이용해 디버그 모드에서는 `preconditionFailure(_:)`, `assertionFailure(_:)`를 통해 런타임 에러 발생 유도, 릴리즈에서는 에러 발생 시 무시

Reply:
출력문의 경우 간단하게 코딩을 할 수 있지만, 단순하게 에러를 프린트만 해주기 때문에 에러의 형식을 특정해줄 수 없으며,형식 안정성이 떨어질 수 있다고 생각했습니다. 2번 방법의 경우에도 throw만 해줄 뿐 크게 다르지 않을 것 같습니다. 이부분은 라이언의 피드백을 바탕으로 찾아보며 공부한 내용을 적은 TIL에 정리해두었습니다!
[2022.01.24. 호랭이 TIL🐯: 에러처리에 Result타입 or 컴파일러 제어문 사용하기](https://camp.yagom-academy.kr/camp/611931356143323c8b376dc3/boards/614c48258a9fd23b353b0734/articles/61eec6d82947bf40f08f3597)

나머지 3번과 4번의 방법중에서, 한번 사용해봐서 조금은 익숙한 Result 타입을 이용한 방법으로 수정해봤습니다. 컴파일러 전처리문을 이용하는 부분은 처음 보는 방식인데, 이부분은 프로젝트가 끝난 뒤 방학 때 시도해보겠습니다!!
Result 타입을 이용하면 아래와 같은 장점이 있다고 생각합니다.
- 에러 형식이 명시적으로 선언된다.
- 타입 캐스팅 없이 에러 처리가 가능하다.
- 형식 추론을 통해 에러 처리 코드가 단순해진다.
- 작업의 결과를 성공과 실패로 명확히 구분 가능하다.
- get 메소드로 에러 처리 코드를 더욱 단순하게 구현할 수 있다.

Reply:
이어서, Result타입의 반환값을 뷰컨트롤러에서 사용할 때 성공 실패 여부에 따라 파싱된 데이터를 반환 또는 에러를 출력하는 메서드를 별도로 만들었습니다. 
기존의 getDecodedData메서드에서는 getParsedData메서드를 호출하여 데이터를 파싱하고 그것을 이용하여 getData메서드 호출로 데이터를 로드하는 역할을 모두 하고 있었는데, 파싱된 데이터를 반환하는 메서드를 getParsedData라는 이름으로 별도로 분리하고 기존의getDecodedData메서드는 명확한 역할을 나타내기 위해서 loadData로 리네이밍 했습니다.